Shared/Samples/Download preplanned map area/DownloadPreplannedMapAreaView.Model.swift

```diff
@@ -0,0 +1,311 @@
+                        )
+                    }
+            }
+        }
+        
+        deinit {
+            // Removes the temporary directory
+            try? FileManager.default.removeItem(at: temporaryDirectory)
+        }
+        
+        /// Updates the displayed map based on the given preplanned map area. If the preplanned map
+        /// area is not nil, the preplanned map area will be downloaded if necessary and updates the map
+        /// to the currently selected preplanned map area. If the preplanned map area is nil, then the map
+        /// is set to the online web map.
+        private func selectedMapDidChange(from oldValue: SelectedMap) {
+            switch selectedMap {
+            case .onlineWebMap:
+                offlineMap = nil
+            case .offlineMap(let info):
+                if info.canDownload {
+                    // If we have not yet downloaded or started downloading, then kick off a
+                    // download and reset selection to prevous selection since we have to download
+                    // the offline map.
+                    selectedMap = oldValue
+                    Task { [weak self] in
+                        await info.download()
+                        self?.updateCanRemoveDownloadedMaps()
+                    }
+                } else if case .success(let mmpk) = info.result {
+                    // If we have already downloaded, then open the map in the mmpk.
+                    offlineMap = mmpk.maps.first
+                } else {
+                    // If we have a failure, then keep the online map selected.
+                    selectedMap = oldValue
+                }
+            }
+        }
+        
+        /// Updates the `canRemoveDownloadedMaps` state.
+        private func updateCanRemoveDownloadedMaps() {
+            canRemoveDownloadedMaps = allOfflineMapModels.contains(where: \.downloadDidSucceed)
+        }
+        
+        /// Cancels all current jobs.
+        private func cancelAllJobs() async {
+            await withTaskGroup(of: Void.self) { group in
+                allOfflineMapModels.forEach { model in
+                    if model.isDownloading {
+                        group.addTask {
+                            await model.cancelDownloading()
+                        }
+                    }
+                }
+            }
+        }
+        
+        // Removes all downloaded maps.
+        func removeDownloadedMaps() {
+            // Sets the current map to the online web map.
+            selectedMap = .onlineWebMap
+            
+            // Update state.
+            canRemoveDownloadedMaps = false
+            
+            Task {
+                // Cancel all current download jobs.
+                await cancelAllJobs()
+                
+                // Remove each download.
+                allOfflineMapModels.forEach { $0.removeDownloadedContent() }
+            }
```

Comment:
This unstructured task is fine because all it does is kick off job cancellation, which is something worth finishing even if the model is deinited.